### DISCUSSION TITLE
Including `computed_field` in JSON schema?

### DISCUSSION THREAD

**Discussion by mgalbier** [pydantic/pydantic#6298] [Question] [answered]
Hi everyone,

I recently switched to Pydantic v2.0b3 to take advantage of some of its new features, including `computed_field`. However, I've noticed that `computed_field` doesn't seem to be included when calling model_json_schema() on a model.

There's a test I found that appears to be set up for testing the schema generation for computed_field, but it's currently being skipped with the comment 'waiting for #4697'. However, upon checking, it seems that issue #4697 is already closed.

I'm using the JSON schemas in my project, and was wondering if there's any information on when `computed_field` will be supported in it.

Thank you in advance for your help!

**Comment by samuelcolvin** [comment]
thanks for reporting, @dmontagu is looking into it.

**Reply by mgalbier** [reply]
Thank you for the quick response!

**Comment by dmontagu** [comment, marked as answer]
@mgalbier There's a difference between `mode='validation'` (which is the default, and doesn't include computed fields) and `mode='serialization'`, which does include them. Here's an example:

```python
from pydantic import BaseModel, computed_field


class Model(BaseModel):
    x: int

    @computed_field
    def y(self) -> int:
        return self.x + 1


print(Model.model_json_schema()))
#> {'properties': {'x': {'title': 'X', 'type': 'integer'}}, 'required': ['x'], 'title': 'Model', 'type': 'object'}

print(Model.model_json_schema(mode='serialization'))
#> {'properties': {'x': {'title': 'X', 'type': 'integer'}, 'y': {'title': 'Y', 'type': 'integer'}}, 'required': ['x', 'y'], 'title': 'Model', 'type': 'object'}
```

I'm not sure what's going on with the reference to 4697 but I'll look into it.

**Reply by mgalbier** [reply]
Thank you for the clarification! I was not aware that there is a separate mode.

Here is the mentioned [test](https://github.com/pydantic/pydantic/blob/70a43f73e6dd5a5036c356e15f6bef59142e256a/tests/test_computed_fields.py#L66C5-L66C37). 
It seems there the `mode` parameter would also be needed on the `model_json_schema()` call.

**Reply by dmontagu** [reply]
@mgalbier thanks for pointing this out. It turns out there were some further improvements we intended to make (documented in that test) that weren't actually implemented yet! But I've implemented them and got that test passing now in https://github.com/pydantic/pydantic/pull/6299

**Reply by jeremander** [reply]
Thanks @dmontagu, I had the exact same question.  Additionally I was curious: why is `y` in your last example above considered a `required` field?  And is there a way to declare it not to be required?
